### ISSUE TITLE
[scroll-animations][web-animations-2] rewinding a finished scroll-driven animation and playing it again

### ISSUE THREAD

**Issue by graouts** [w3c/csswg-drafts#11270]
I'm working on the WebKit implementation of Scroll-driven Animations and am having issues making the "Animation finish event is fired again after replaying from start" subtest of [scroll-animations/scroll-timelines/updating-the-finished-state.html](https://github.com/web-platform-tests/wpt/blob/master/scroll-animations/scroll-timelines/updating-the-finished-state.html) work. The goal of that test is to "scrub" a scroll-driven animation to its finished state and then reset it back to its 0% time to play again.

Here's a slightly reduced self-contained test with some logging to query the timeline and animation state in that case:

```html
<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<title></title>
</head>
<body style="height: 200vh">
<div></div>
<script>

(async function () {
    const timeline = new ScrollTimeline({ source: document.documentElement });
    const animation = document.querySelector("div").animate({ "marginLeft": "100px" }, { timeline });

    const log = msg => {
        console.log(`${msg}\ntimeline time = ${timeline.currentTime}, animation time = ${animation.currentTime}, animation start time = ${animation.startTime}`);
    };

    await animation.ready;
    log("1. After animation readiness");

    const maxScroll = timeline.source.scrollHeight - timeline.source.clientHeight;
    timeline.source.scrollTop = maxScroll;
    await new Promise(requestAnimationFrame);
    log("2. After scrolling to 100% and waiting one frame");

    timeline.source.scrollTop = 0;
    animation.play();
    log("3. After scrolling back to 0% and calling play()");

    await new Promise(requestAnimationFrame);
    log("4. After animation readiness");
})();

</script>  
</body>
</html>
```

**Chrome Canary**
```
1. After animation readiness
timeline time = 0%, animation time = 0%, animation start time = 0%

2. After scrolling to 100% and waiting one frame
timeline time = 100%, animation time = 100%, animation start time = 0%

3. After scrolling back to 0% and calling play()
timeline time = 100%, animation time = 100%, animation start time = null

4. After animation readiness
timeline time = 0%, animation time = 0%, animation start time = 0%
```

**WebKit**
```
1. After animation readiness
timeline time = 0%, animation time = 0%, animation start time = 0%

2. After scrolling to 100% and waiting one frame
timeline time = 100%, animation time = 100%, animation start time = 0%

3. After scrolling back to 0% and calling play()
timeline time = 100%, animation time = 100%, animation start time = null

4. After animation readiness
timeline time = 0%, animation time = 100%, animation start time = -100%
```

The logging is identical in Chrome Canary and WebKit up until we rewind the animation and play it again. The result of what happens in Chrome makes _complete_ sense, but I'm reading the spec over and over and failing to see what should be happening to match the Chrome behavior.

The internal state as `play()` is called is as follows:

```
hold time = 100%
start time = unresolved
```

And then as the pending play task is performed:

```
ready time = 0%
start time = ready time - hold time (100%) = -100%
hold time = unresolved
```

Can someone read the spec in a way that yields a different result for the values seen during the performance of the pending play task?

**Comment by graouts** [comment]
Cc the Chrome folks knowledgeable in this area: @flackr, @kevers-google and @andruud.

**Comment by ydaniv** [comment]
@graouts why do you need to call `animation.play()` again?

**Comment by graouts** [comment]
This is just how the WPT test that I'm trying to understand works.

**Comment by graouts** [comment]
Does Chrome perform the pending play task prior to updating the current time on the timeline? This would make the `ready time` be `100%` and as a result the `start time` be `0%`.

**Comment by kevers-google** [comment]
From looking at the log output, it appears that Chrome recomputed a start time, without the constraint that it needed to either preserve current-time (to match the hold-time and avoid a jump) or to auto-rewind. Both of these constraints only makes sense only for time-based animations.

**Comment by graouts** [comment]
Oh, so Chrome seems to expand step 7 in this way (bold italic is the added text):

> If has finite timeline and **_either_** previous current time is unresolved **_or the auto-rewind flag is true_**:
> - Set the flag auto align start time to true.

**Comment by graouts** [comment]
I think this bit of code in [Chrome's `Animation::UpdateFinishedState()`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/animation/animation.cc;l=1797;drc=277f4ab48eb85f7441f78aed191c31068ce89814)  is also relevant to the WPT test, though not the reduced test I posted:

```c++
  // TODO(kevers): Add a new step to the spec.
  // Clear finished state and abort the procedure if play-pending and waiting
  // for a new start time.
  if (timeline_ && timeline_->IsScrollTimeline() && pending_play_ &&
      auto_align_start_time_) {
    finished_ = false;
    pending_finish_notification_ = false;
    committed_finish_notification_ = false;
    return;
  }
```

**Comment by kevers-google** [comment]
If I recall correctly, this batch of tests was added prior to the introduction of auto-aligned start times, and was essentially a duplicate of equivalent tests with a document time.  The Blink implementation has changed considerably since the test was introduced.   The intent of the test was to verify that onfinish would fire each time we scroll to the end. I believe the test should read:

```
animation.onfinish = event => {
  window.requestAnimationFrame(() => {
    scroller.scrollTop = 0;
    window.requestAnimationFrame(() => {
      scroller.scrollTop = maxScroll;
      animation.onfinish = ...
    });
  });
}
```
The call to play() is not needed if we ensure that we have an animation tick with the reset scroll position.

This test is passing in Blink for another reason (likely requiring a spec tweak).  In Blink, explicit calls to play() reset any stickiness of the start time and force a new start time to be evaluated.  Stickiness in the timeline arises when explicitly setting the currentTime of a scroll-driven animation. This overrides the auto-aligned start time, at while point the animation progress is no longer strictly aligned with the calculated range.  In this example, the start time is not sticky, and it is arguable if play should simply abort as a no-op.  In the Blink implementation, any call to play with a scrollTimeline forces a new startTime to be evaluated.

As the test is currently written, play() is indirectly triggering the reset in UpdateFinishedState.  The reason for the change to UpdateFinishedState was that we cannot update the finished state while waiting on an deferred start time.  We should have a separate test for an explicit call to play resetting stickiness. I believe one already exists, but have not hunted it down.
